### PR TITLE
Add OIDC issuer validation

### PR DESCRIPTION
**Pull request by Avery-Dunn**
Adds new validation according to: https://identitydivision.visualstudio.com/Engineering/_workitems/edit/3268768

The issuer returned by the call to the OIDC endpoint (.well-known/openid-configuration) is valid if one of the following is true:
- It exactly matches the authority URL
- It has a known Microsoft host (e.g., login.microsoftonline.com)
- It has the same scheme and host as the authority (path can be different)
- For CIAM, the issuer follows the pattern of {tenant}.ciamlogin.com, even when using a custom domain

The validation behavior was added to `authority.py`, and new tests were added to `test_authority.py` to cover it

In addition, a few other tests in `test_authority.py` and  `test_application.py` were given valid issuers in their mocked OIDC discovery responses in order to pass validation, but were otherwise unchanged.

(this is the full implementation of the work started in this draft PR: https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/830)

### CODE REVIEW

File: msal/authority.py [hunk 1]
```diff
@@ -95,11 +96,19 @@ def __init__(
...
+        self._issuer = openid_config.get('issuer')
         self.authorization_endpoint = openid_config['authorization_endpoint']
         self.token_endpoint = openid_config['token_endpoint']
         self.device_authorization_endpoint = openid_config.get('device_authorization_endpoint')
         _, _, self.tenant = canonicalize(self.token_endpoint)  # Usually a GUID
 
+        # Validate the issuer if using OIDC authority
+        if self._oidc_authority_url and not self.has_valid_issuer():
+            raise ValueError((
+                "The issuer '{iss}' does not match the authority '{auth}' or a known pattern. "
```
Comment (bgavrilMS) [inline msal/authority.py:131]: Maybe add a note that for Entra, they should use `authority_url`

**Review by bgavrilMS — APPROVED** [review]

**Review by rayluo — CHANGES_REQUESTED** [review]
We are discussing in this [offline conversation on Teams](https://teams.microsoft.com/l/message/19:34edbc40f791488eac69171721f6eef2@thread.tacv2/1753203379130?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=f2d72515-5094-4927-915d-edc363be6fc3&parentMessageId=1753121351714&teamName=MSAL%20Confidential%20Client&channelName=Daily%20Standup&createdTime=1753203379130).

File: msal/authority.py (hunk 1 above)
Comment (Avery-Dunn) [reply msal/authority.py:131]: In the latest commit it's now more clear that the regular authority API should be used when you have an Entra authority

File: tests/test_authority.py [hunk 2]
```diff
@@ -115,25 +116,37 @@ def setUp(self):
             # Here the test is to confirm the OIDC endpoint contains no "/v2.0"
             self.authority + "/.well-known/openid-configuration")
 
+    def setup_tenant_discovery(self, tenant_discovery):
```
Comment (bgavrilMS) [inline tests/test_authority.py:120]: Is acceptance test 2 from Ray's list covered?

Acceptance test case 2: MSAL shall error out
contoso.com/.well-known/openid-configuration returns {"issuer": "login.microsoftonline.com/tenant", ...}

Comment (Avery-Dunn) [reply tests/test_authority.py:120]: Yes, that's handled by the test called "test_custom_authority_with_microsoft_issuer"

It uses a custom authority and a login.microsoftonline.com issuer and checks for the validation exception

**Review by bgavrilMS — APPROVED** [review]

**Review by bgavrilMS — APPROVED** [review]

**Comment by bgavrilMS** [comment]
@rayluo - any updates on this?

**Comment by Avery-Dunn** [comment]
@rayluo Any updates on this? A while back you had some behaviors you wanted to confirm or edge cases you wanted to test, and @4gust updated this PR to handle validating region-specific endpoints, but I can't remember what was still blocking this.

File: msal/authority.py [hunk 3]
```diff
@@ -21,6 +21,26 @@
     'login-us.microsoftonline.com',
     AZURE_US_GOVERNMENT,
     ])
+
+# Trusted issuer hosts for OIDC issuer validation
+# Includes all well-known Microsoft identity provider hosts and national clouds
+TRUSTED_ISSUER_HOSTS = frozenset([
```
Comment (bgavrilMS) [inline msal/authority.py:27]: Let's please add the 3 new clouds here too: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/5709/changes

File: msal/authority.py [hunk 4]
```diff
@@ -175,6 +206,60 @@ def user_realm_discovery(self, username, correlation_id=None, response=None):
...
+    def has_valid_issuer(self):
+        """
+        Returns True if the issuer from OIDC discovery is valid for this authority.
+
+        An issuer is valid if one of the following is true:
+        - It exactly matches the authority URL (with/without trailing slash)
+        - It has the same scheme and host as the authority (path can be different)
+        - The issuer host is a well-known Microsoft authority host
+        - The issuer host is a regional variant of a well-known host (e.g., westus2.login.microsoft.com)
+        - For CIAM, the issuer follows the pattern of {tenant}.ciamlogin.com
```
Comment (bgavrilMS) [inline msal/authority.py]: Did we do smth for B2C as well?

File: msal/authority.py [hunk 5]
```diff
@@ -175,6 +206,60 @@ def user_realm_discovery(self, username, correlation_id=None, response=None):
...
 
+    def has_valid_issuer(self):
+        """
+        Returns True if the issuer from OIDC discovery is valid for this authority.
+
+        An issuer is valid if one of the following is true:
+        - It exactly matches the authority URL (with/without trailing slash)
+        - It has the same scheme and host as the authority (path can be different)
+        - The issuer host is a well-known Microsoft authority host
+        - The issuer host is a regional variant of a well-known host (e.g., westus2.login.microsoft.com)
```
Comment (bgavrilMS) [inline msal/authority.py:220]: Don't we need a regional rule for not well-known hosts? 

authority: `region.someothercloud.com` issuer: someothercloud.com

Expected: ok
Actual: ?

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
Need another rule for regiona.

File: msal/authority.py (hunk 4 above)
Comment (4gust) [reply msal/authority.py]: Updated this to check for well knows B2C URL's

File: msal/authority.py (hunk 5 above)
Comment (4gust) [reply msal/authority.py:220]: Updated this.

File: msal/authority.py (hunk 3 above)
Comment (4gust) [reply msal/authority.py:27]: Added the new clouds

**Review by bgavrilMS — APPROVED** [review]
